### ISSUE TITLE
when I closed the Producer, I found that the memory was not freed

### ISSUE THREAD

**Issue by jessonliu** [versatica/libmediasoupclient#69]
when I closed the Producer, I found that the memory was not freed!
I looked at the source code and found a problem!
the "void Sdp::RemoteSdp::DisableMediaSection" function is called  in "void SendHandler::StopSending(const std::string& localId)" implementation;
DisableMediaSection function Modify remoteSdp, However, the remoteSdp Is still recvonly;

**Comment by ibc** [comment]
Where exactly is the memory leak and how is that related to "sendonly" at all?

**Comment by jessonliu** [comment]
> 内存泄漏到底在哪里？它与“仅发送”有什么关系？

**Comment by jessonliu** [comment]
> Where exactly is the memory leak and how is that related to "sendonly" at all?
Class RemoteSdp
Function
void Sdp::RemoteSdp::DisableMediaSection(const std::string& mid)
{
	MSC_TRACE();

	// TODO: Should check that mediaSections.find(mid) exists.

	auto* mediaSection = this->mediaSections[mid];

	mediaSection->Disable();
}

mediaSection->Disable(); Failed to modify SDP

**Comment by ibc** [comment]
Where is the memory leak?

Why have you closed the issue?

**Comment by jessonliu** [comment]
> Where exactly is the memory leak and how is that related to "sendonly" at all?

About memory don't release, because there is no implementation RemoteSdp: : CloseMediaSection function.
When I create multiple producers and then close the producers, the memory is not freed

**Comment by jessonliu** [comment]
> Where is the memory leak?
> 
> Why have you closed the issue?

I use libmediasoupclient to bridge object-c and make mediasoup for iOS!
When sharing the screen on iOS, we need to create a Producer. When ending the screen sharing, we need to close the Producer! I found that memory is not freed after repeated operations! So I view source found SDP m line will increase continuously, then I shall, in accordance with the js version implements the void SDP: : RemoteSdp: : CloseMediaSection (const STD: : string & mid) such as function, then repeat sharing screen, there are created in the producer will increase, when close the producer will fall!

**Comment by ibc** [comment]
The remote SDP grows indefinitely because closed closed m= sections are not reused, but just closed (with media port set to 0) and another m= section is created for any new Producer. However:

1. We have pending changes in libmediasoupclient to reuse closed m= sections.
2. Even if we don't do bullet 1, we should indeed `delete` the closed `MediaSection`.

**Comment by jessonliu** [comment]
> The remote SDP grows indefinitely because closed closed m= sections are not reused, but just closed (with media port set to 0) and another m= section is created for any new Producer. However:
> 
> 1. We have pending changes in libmediasoupclient to reuse closed m= sections.
> 2. Even if we don't do bullet 1, we should indeed `delete` the closed `MediaSection`.

Thank you for your reply, perhaps the recvonly modification failure problem mentioned above is not clear! So I took a screenshot and marked it, please have a look, I really found this problem in the development! Hope to contribute to mediasoup!

* https://github.com/jessonliu/PictureFile/blob/master/picture.jpg

```c++
void Sdp::RemoteSdp::AddMediaSection(MediaSection* mediaSection)
{
	MSC_TRACE();

	// Store it in the map.
	this->mediaSections[mediaSection->GetMid()] = mediaSection;

	// Update SDP object.
	this->sdpObject["media"].push_back(mediaSection->GetObject());

	std::string mids = this->sdpObject["groups"][0]["mids"].get<std::string>();
	if (mids.empty())
		mids = mediaSection->GetMid();
	else
		mids.append(" ").append(mediaSection->GetMid());

	this->sdpObject["groups"][0]["mids"] = mids;
}
```

this->sdpObject saves the "mediaSection->GetObject()",  is no the object MediaSection

**Comment by jessonliu** [comment]
`
void SendHandler::StopSending(const std::string& localId)
{
	MSC_TRACE();

	MSC_DEBUG("[localId:%s]", localId.c_str());

	auto jsonLocaIdIt = this->mapMidTransceiver.find(localId);
	if (jsonLocaIdIt == this->mapMidTransceiver.end())
		throw Exception("Associated RtpTransceiver not found");

	auto transceiver = jsonLocaIdIt->second;

	transceiver->sender()->SetTrack(nullptr);
	this->pc->RemoveTrack(transceiver->sender());
	this->remoteSdp->CloseMediaSection(transceiver->mid().value());

	// May throw.
	webrtc::PeerConnectionInterface::RTCOfferAnswerOptions options;

	auto offer = this->pc->CreateOffer(options);
    
//    std::cout << "🌹🌹🌹🌹offer:\n" <<  offer.c_str() << std::endl;

	MSC_DEBUG("calling pc->SetLocalDescription() [offer:%s]", offer.c_str());

	// May throw.
	this->pc->SetLocalDescription(PeerConnection::SdpType::OFFER, offer);

	auto localSdpObj = sdptransform::parse(this->pc->GetLocalDescription());
	auto answer      = this->remoteSdp->GetSdp();
    
    std::cout << "🌹🌹🌹🌹answer:/n" <<  answer.c_str() << std::endl;

	MSC_DEBUG("calling pc->SetRemoteDescription() [answer:%s]", answer.c_str());

	// May throw.
	this->pc->SetRemoteDescription(PeerConnection::SdpType::ANSWER, answer);
}
`

**Comment by ibc** [comment]
Yes, I see. Indeed a bug. Will be fixed in next version.

**Comment by ibc** [comment]
Please, do not close the issue. It's not fixed.

**Comment by jmillan** [comment]
This has been fixed b219aa51da32c07742ee4ca86a71245e88d60012
